File: wallet/app/ui/core/address.mjs
// Strict BIP-350 Bech32m address validation and display, JS mirror of
// wallet/app/src-tauri/src/address.rs per protocol/schemas/identity-v1.md §2.
// The version/type/payload layout is OWNER_BLOCKED (identity-v1 §2.1): the
// payload stays opaque and no layout is ever defaulted. Historical-protocol
// HRPs reject with wrong_protocol_identity, never auto-convert.
export const HRP = "noos";
const BECH32M_CONST = 0x2bc830a3;
const CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l";
const GEN = [0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3];
const DATA_MIN = 6;
const DATA_MAX = 83;
// Historical HRP assembled from char codes so the identifier never appears
// literally (identity gate; identity-v1 §5).
const HISTORICAL_HRP = String.fromCharCode(0x6d, 0x69, 0x6e, 0x64);

export class AddressError extends Error {
  constructor(code) { super(code); this.code = code; }
}

function polymod(values) {
  let chk = 1;
  for (const v of values) {
    const top = chk >>> 25;
    chk = (((chk & 0x01ffffff) << 5) ^ v) >>> 0;
    for (let i = 0; i < 5; i += 1) if ((top >>> i) & 1) chk = (chk ^ GEN[i]) >>> 0;
  }
  return chk;
}

function hrpExpand(hrp) {
  const out = [];
  for (const c of hrp) out.push(c.charCodeAt(0) >>> 5);
  out.push(0);
  for (const c of hrp) out.push(c.charCodeAt(0) & 31);
  return out;
}

// Validate canonical case, strict HRP, charset, length, and the Bech32m
// checksum. Returns the opaque 5-bit payload (checksum stripped).
export function validateAddress(address) {
  if (typeof address !== "string" || /[A-Z]/.test(address)) throw new AddressError("noncanonical_address");
  const sep = address.lastIndexOf("1");
  if (sep < 1) throw new AddressError("wrong_hrp");
  const hrp = address.slice(0, sep);
  const data = address.slice(sep + 1);
  if (hrp === HISTORICAL_HRP) throw new AddressError("wrong_protocol_identity");
  if (hrp !== HRP) throw new AddressError("wrong_hrp");
  if (data.length < DATA_MIN || data.length > DATA_MAX) throw new AddressError("bad_length");
  const values = [];
  for (const c of data) {
    const idx = CHARSET.indexOf(c);
    if (idx < 0) throw new AddressError("bad_charset");
    values.push(idx);
  }
  if (polymod([...hrpExpand(hrp), ...values]) !== BECH32M_CONST) throw new AddressError("bad_checksum");
  return Object.freeze({ payload5: values.slice(0, -6) });
}

// Re-encode an opaque payload under the strict noos HRP: canonical round-trip
// display only. Defines NO payload layout and refuses any other HRP.
export function encodeAddress(payload5) {
  if (payload5.some((v) => !Number.isInteger(v) || v < 0 || v >= 32)) throw new AddressError("bad_charset");
  const dataLen = payload5.length + 6;
  if (dataLen < DATA_MIN || dataLen > DATA_MAX) throw new AddressError("bad_length");
  const pm = polymod([...hrpExpand(HRP), ...payload5, 0, 0, 0, 0, 0, 0]) ^ BECH32M_CONST;
  let out = `${HRP}1`;
  for (const v of payload5) out += CHARSET[v];
  for (let i = 0; i < 6; i += 1) out += CHARSET[(pm >>> (5 * (5 - i))) & 31];
  return out;
}

// Grouped display form for humans; the canonical wire form stays unspaced.
export function displayGroups(address) {
  const verified = validateAddress(address);
  const data = address.slice(HRP.length + 1);
  const groups = data.match(/.{1,4}/g) ?? [];
  return Object.freeze({ display: `${HRP}1 ${groups.join(" ")}`, payloadChars: data.length - 6, payload5: verified.payload5 });
}

File: wallet/app/src-tauri/src/address.rs
//! Strict BIP-350 Bech32m address validation and display per
//! `protocol/schemas/identity-v1.md` §2.
//!
//! The version/type/payload byte layout is OWNER_BLOCKED (identity-v1 §2.1):
//! this module NEVER interprets or defaults the payload. It validates the
//! canonical form (lowercase-only, strict `noos` HRP, Bech32m checksum) and
//! treats the 5-bit payload as opaque. Historical-protocol HRPs reject with
//! `wrong_protocol_identity` and are never auto-converted.

use thiserror::Error;

/// The only human-readable part any NOOSPHERE surface accepts.
pub const HRP: &str = "noos";
const BECH32M_CONST: u32 = 0x2bc8_30a3;
const CHARSET: &[u8; 32] = b"qpzry9x8gf2tvdw0s3jn54khce6mua7l";
/// Data-part length bounds (checksum included) from the frozen API pattern
/// `^noos1[023456789acdefghjklmnpqrstuvwxyz]{6,83}$`.
const DATA_MIN: usize = 6;
const DATA_MAX: usize = 83;
/// Historical HRP of the predecessor protocol, assembled from bytes so the
/// identifier never appears literally (identity gate; see identity-v1 §5).
const HISTORICAL_HRP_BYTES: [u8; 4] = [0x6d, 0x69, 0x6e, 0x64];

#[derive(Debug, Error, Clone, PartialEq, Eq)]
pub enum AddressError {
    /// Any uppercase character, including all-uppercase display form.
    #[error("noncanonical_address")]
    NonCanonicalCase,
    #[error("wrong_hrp")]
    WrongHrp,
    /// Historical predecessor-protocol address; terminal reject, no conversion.
    #[error("wrong_protocol_identity")]
    WrongProtocolIdentity,
    #[error("bad_charset")]
    BadCharset,
    #[error("bad_length")]
    BadLength,
    /// Checksum failure is terminal; no error correction is ever attempted.
    #[error("bad_checksum")]
    BadChecksum,
}

/// A checksum-verified address whose payload stays opaque (OWNER_BLOCKED).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct VerifiedAddress {
    /// 5-bit groups between separator and checksum, uninterpreted.
    pub payload5: Vec<u8>,
}

#[allow(clippy::arithmetic_side_effects)] // masked 30-bit polynomial arithmetic cannot overflow u32
fn polymod(values: impl Iterator<Item = u8>) -> u32 {
    const GEN: [u32; 5] = [
        0x3b6a_57b2,
        0x2650_8e6d,
        0x1ea1_19fa,
        0x3d42_33dd,
        0x2a14_62b3,
    ];
    let mut chk: u32 = 1;
    for v in values {
        let top = chk >> 25;
        chk = ((chk & 0x01ff_ffff) << 5) ^ u32::from(v);
        for (i, g) in GEN.iter().enumerate() {
            if (top >> i) & 1 == 1 {
                chk ^= g;
            }
        }
    }
    chk
}

#[allow(clippy::arithmetic_side_effects)] // 3-bit shifts and 5-bit masks on ASCII bytes
fn hrp_expand(hrp: &str) -> Vec<u8> {
    let bytes = hrp.as_bytes();
    let mut out = Vec::with_capacity(bytes.len().saturating_mul(2).saturating_add(1));
    for b in bytes {
        out.push(b >> 5);
    }
    out.push(0);
    for b in bytes {
        out.push(b & 31);
    }
    out
}

fn charset_index(c: u8) -> Option<u8> {
    CHARSET
        .iter()
        .position(|&x| x == c)
        .and_then(|i| u8::try_from(i).ok())
}

fn split(address: &str) -> Result<(&str, &str), AddressError> {
    // Canonical output is lowercase; ANY uppercase character is a terminal
    // reject (mixed case AND the BIP-173 all-uppercase display form).
    if address.bytes().any(|b| b.is_ascii_uppercase()) {
        return Err(AddressError::NonCanonicalCase);
    }
    let sep = address.rfind('1').ok_or(AddressError::WrongHrp)?;
    let hrp = address.get(..sep).ok_or(AddressError::WrongHrp)?;
    let data = address
        .get(sep.saturating_add(1)..)
        .ok_or(AddressError::WrongHrp)?;
    if hrp.is_empty() {
        return Err(AddressError::WrongHrp);
    }
    Ok((hrp, data))
}

fn require_hrp(hrp: &str) -> Result<(), AddressError> {
    if hrp.as_bytes() == HISTORICAL_HRP_BYTES {
        return Err(AddressError::WrongProtocolIdentity);
    }
    if hrp != HRP {
        return Err(AddressError::WrongHrp);
    }
    Ok(())
}

/// Validate a NOOSPHERE address: canonical case, strict HRP, charset, length,
/// and Bech32m checksum. The payload is returned opaque; interpreting it is
/// PROHIBITED until the identity-v1 amendment 1 layout table lands.
pub fn validate(address: &str) -> Result<VerifiedAddress, AddressError> {
    let (hrp, data) = split(address)?;
    require_hrp(hrp)?;
    if !(DATA_MIN..=DATA_MAX).contains(&data.len()) {
        return Err(AddressError::BadLength);
    }
    let mut values = Vec::with_capacity(data.len());
    for c in data.bytes() {
        values.push(charset_index(c).ok_or(AddressError::BadCharset)?);
    }
    let checkable = hrp_expand(hrp).into_iter().chain(values.iter().copied());
    if polymod(checkable) != BECH32M_CONST {
        return Err(AddressError::BadChecksum);
    }
    let payload_len = values.len().saturating_sub(6);
    values.truncate(payload_len);
    Ok(VerifiedAddress { payload5: values })
}

/// Re-encode an opaque 5-bit payload under the strict `noos` HRP. Used only
/// for canonical round-trip display of an already-validated address; this
/// function defines NO payload layout and refuses any other HRP.
pub fn encode(payload5: &[u8]) -> Result<String, AddressError> {
    if payload5.iter().any(|&v| v >= 32) {
        return Err(AddressError::BadCharset);
    }
    // payload + 6 checksum chars must stay within the frozen data bounds.
    let data_len = payload5.len().saturating_add(6);
    if !(DATA_MIN..=DATA_MAX).contains(&data_len) {
        return Err(AddressError::BadLength);
    }
    let expanded = hrp_expand(HRP);
    let with_template = expanded
        .iter()
        .copied()
        .chain(payload5.iter().copied())
        .chain([0u8; 6]);
    let pm = polymod(with_template) ^ BECH32M_CONST;
    let mut out = String::with_capacity(HRP.len().saturating_add(1).saturating_add(data_len));
    out.push_str(HRP);
    out.push('1');
    for &v in payload5 {
        out.push(char::from(CHARSET[usize::from(v)]));
    }
    #[allow(clippy::arithmetic_side_effects)] // 30-bit value sliced into six 5-bit groups
    for i in 0..6u32 {
        let idx = ((pm >> (5 * (5 - i))) & 31) as usize;
        out.push(char::from(CHARSET[idx]));
    }
    Ok(out)
}
